### PR TITLE
#161255193 allow user to signup 

### PR DESCRIPTION
**Pull request by Mirriam-Maina**
### What does this PR do?
Allows users should be to Signup and or Register using Sign up forms on the signup route.

### Description of Task to be completed?
- Signup a user
- Validation of inputs
- Show validation errors
- Redirect user upon successful signup
- Show signup success message

### How should this be manually tested?
- Access the link (https://ah-maps-frontend-staging.herokuapp.com/signup)
- Sign up a user

### What are the relevant pivotal tracker stories?
[Users should be able Signup/Register to Authors Haven](https://www.pivotaltracker.com/story/show/161255193)

### Screenshots (if appropriate)
screenshot 2018-12-05 at 10 31 59
screenshot 2018-12-05 at 10 34 26

### Checklist:
- [x] My code follows the style guidelines of this project
- [ ] My teammates have reviewed my PR
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] My PR has one commit.

### CODE REVIEW

File: src/containers/SignUp/index.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      register({ ...this.state });
+    }
+  }
+
+  render() {
+    const { signup } = this.props;
+    const {
+      username, password, email, confirmPassword, errors,
+    } = this.state;
+    const inputs = [
```
Comment (Rodgers-M) [inline src/containers/SignUp/index.js:80]: the only thing that changes here is the label name, can it be dynamic? with that then this array can be reduced to one object

**Review by Rodgers-M — COMMENTED** [review]
please see comments in @mmosoroohh 's PR on some similar issues you have like the api tokens and also abracting some of the `jsx`

File: src/containers/SignUp/index.js (hunk 1 above)
Comment (p8ul) [reply src/containers/SignUp/index.js:80]: Thanks for the feedback. The object has other changes besides the label name, the errors are also dynamic from the current state. For example, each input item has errors that are specific to it. The value of the objects also changes dynamically from the state.

File: src/constants/auth.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+export const TOKEN_KEY = 'jwtToken';
+export const USER_TOKEN_TEST = {
+  user: {
+    email: 'map@gmail.com',
```
Comment (Rodgers-M) [inline src/constants/auth.js:4]: this email and username should also be in the environment

File: src/redux/actions/signup.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  payload,
+});
+
+export const SignUpRequest = data => (dispatch) => {
+  api({ url: 'users/', method: 'post', data })
+    .then((response) => {
+      const res = { message: response.data.user.success };
+      dispatch(signUpSuccess(res));
+      history.replace('/login');
+      setTimeout(() => { window.location.reload(); }, 3000);
```
Comment (Rodgers-M) [inline src/redux/actions/signup.js]: any reason that the browser needs to reload here ?

**Pull request merged**